File: web/src/utils/queryClassify.js
/*-------------------------------------------------------------------------
 *
 * pgEdge Natural Language Agent
 *
 * Copyright (c) 2025 - 2026, pgEdge, Inc.
 * This software is released under The PostgreSQL License
 *
 *-------------------------------------------------------------------------
 */

const WRITE_PREFIXES = [
    'CREATE', 'DROP', 'ALTER', 'TRUNCATE',
    'INSERT', 'UPDATE', 'DELETE',
];

const READ_PREFIXES = [
    'SELECT', 'WITH', 'TABLE', 'VALUES',
    'EXPLAIN', 'SHOW',
];

/**
 * Classifies whether a SQL query is a write (DDL/DML) operation.
 * Read queries (SELECT, WITH, etc.) return false.
 * Write queries (CREATE, DROP, INSERT, etc.) return true.
 * Unknown query types are treated as potentially destructive.
 *
 * @param {string} sql - The SQL query to classify
 * @returns {boolean} - True if the query is a write operation
 */
export function isWriteQuery(sql) {
    if (!sql || typeof sql !== 'string') return false;
    const upper = sql.trim().toUpperCase();
    if (READ_PREFIXES.some(p => upper.startsWith(p))) return false;
    if (WRITE_PREFIXES.some(p => upper.startsWith(p))) return true;
    // Unknown query types are treated as potentially destructive
    return true;
}

File: internal/chat/query_classify.go
/*-------------------------------------------------------------------------
 *
 * pgEdge Natural Language Agent
 *
 * Copyright (c) 2025 - 2026, pgEdge, Inc.
 * This software is released under The PostgreSQL License
 *
 *-------------------------------------------------------------------------
 */

package chat

import "strings"

// QueryType represents the type of SQL query
type QueryType int

const (
	QueryTypeSelect QueryType = iota
	QueryTypeDDL
	QueryTypeDML
	QueryTypeOther
)

// ClassifyQuery determines if a SQL query is a read or write
// operation. Returns the query type and whether the query modifies
// data.
func ClassifyQuery(sql string) (QueryType, bool) {
	upper := strings.ToUpper(strings.TrimSpace(sql))

	switch {
	case strings.HasPrefix(upper, "SELECT"),
		strings.HasPrefix(upper, "WITH"),
		strings.HasPrefix(upper, "TABLE"),
		strings.HasPrefix(upper, "VALUES"),
		strings.HasPrefix(upper, "EXPLAIN"),
		strings.HasPrefix(upper, "SHOW"):
		return QueryTypeSelect, false

	case strings.HasPrefix(upper, "CREATE"),
		strings.HasPrefix(upper, "DROP"),
		strings.HasPrefix(upper, "ALTER"),
		strings.HasPrefix(upper, "TRUNCATE"):
		return QueryTypeDDL, true

	case strings.HasPrefix(upper, "INSERT"),
		strings.HasPrefix(upper, "UPDATE"),
		strings.HasPrefix(upper, "DELETE"):
		return QueryTypeDML, true

	default:
		return QueryTypeOther, true
	}
}
